packages/fdc3-agent-proxy/src/listeners/PrivateChannelEventListener.ts

```diff
@@ -1,62 +1,158 @@
+          type = 'disconnect';
+          details = null;
+          break;
+      }
+
+      const event: PrivateChannelEvent = {
+        type,
+        details,
+      };
+      handler(event);
+    };
+
+    super(
+      messaging,
+      channelId,
+      [
+        'privateChannelOnAddContextListenerEvent',
+        'privateChannelOnUnsubscribeEvent',
+        'privateChannelOnDisconnectEvent',
+      ],
+      'addContextListener',
+      wrappedHandler
+    );
   }
+}
+
+export class PrivateChannelDisconnectEventListener extends AbstractPrivateChannelEventListener {
+  constructor(messaging: Messaging, channelId: string, handler: EventHandler) {
+    const wrappedHandler = (msg: PrivateChannelEventMessages) => {
+      if (isPrivateChannelOnDisconnectEvent(msg)) {
+        const event: PrivateChannelDisconnectEvent = {
+          type: 'disconnect',
+          details: null,
+        };
+        handler(event);
+      } else {
+        console.error('PrivateChannelDisconnectEventListener was called for a different message type!', msg);
+      }
+    };
 
-  action(_m: any): void {
-    this.handler();
+    super(messaging, channelId, ['privateChannelOnDisconnectEvent'], 'disconnect', wrappedHandler);
   }
 }
 
-export class PrivateChannelEventListenerType extends AbstractPrivateChannelEventListener<(m: string) => void> {
-  constructor(
-    messaging: Messaging,
-    channelId: string,
-    listenerType: EVENT_TYPES_WITH_TYPE_HANDLER,
-    handler: (s: string) => void
-  ) {
-    super(messaging, channelId, listenerType, handler);
+export class PrivateChannelAddContextEventListener extends AbstractPrivateChannelEventListener {
+  constructor(messaging: Messaging, channelId: string, handler: EventHandler) {
+    const wrappedHandler = (msg: PrivateChannelEventMessages) => {
+      if (isPrivateChannelOnAddContextListenerEvent(msg)) {
+        const event: ApiEvent = {
+          type: 'addContextListener',
+          details: { contextType: msg.payload.contextType },
+        };
+        handler(event);
+      } else {
+        console.error('PrivateChannelDisconnectEventListener was called for a different message type!', msg);
+      }
+    };
+    super(messaging, channelId, ['privateChannelOnAddContextListenerEvent'], 'addContextListener', wrappedHandler);
   }
+}
 
-  action(m: any): void {
-    this.handler(m.payload.contextType);
+export class PrivateChannelUnsubscribeEventListener extends AbstractPrivateChannelEventListener {
+  constructor(messaging: Messaging, channelId: string, handler: EventHandler) {
+    const wrappedHandler = (msg: PrivateChannelEventMessages) => {
+      if (isPrivateChannelOnUnsubscribeEvent(msg)) {
+        const event: ApiEvent = {
+          type: 'unsubscribe',
+          details: { contextType: msg.payload.contextType },
+        };
+        handler(event);
+      } else {
+        console.error('PrivateChannelDisconnectEventListener was called for a different message type!', msg);
```

Comment:
fixed the message - have not added a test, its just single log line. I do want it to create noise if this starts happening (rather than it happening silently) but ideally it never happens.